File: data/proofs/test_dashboard.py
#!/usr/bin/env python3
# test_dashboard.py - Headless browser test for zkPrologML dashboard

from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.options import Options
import time

def test_dashboard():
    print("🔮 Testing zkPrologML Dashboard")
    print("=" * 60)
    
    # Setup headless Chrome
    chrome_options = Options()
    chrome_options.add_argument("--headless")
    chrome_options.add_argument("--no-sandbox")
    chrome_options.add_argument("--disable-dev-shm-usage")
    
    driver = webdriver.Chrome(options=chrome_options)
    
    try:
        # Load dashboard
        url = "http://localhost/"
        print(f"\n📥 Loading: {url}")
        driver.get(url)
        
        # Wait for page to load
        time.sleep(2)
        
        # Test 1: Check title
        print("\n✓ Test 1: Page title")
        title = driver.title
        print(f"  Title: {title}")
        assert "zkPrologML" in title, "Title should contain 'zkPrologML'"
        
        # Test 2: Check stats are visible
        print("\n✓ Test 2: Stats cards")
        stats = driver.find_elements(By.CLASS_NAME, "card")
        print(f"  Found {len(stats)} stat cards")
        assert len(stats) >= 6, "Should have at least 6 stat cards"
        
        # Test 3: Check Frank is present
        print("\n✓ Test 3: Frank assistant")
        frank_input = driver.find_element(By.ID, "prologInput")
        print(f"  Frank input placeholder: {frank_input.get_attribute('placeholder')}")
        assert frank_input is not None, "Frank input should exist"
        
        # Test 4: Check error log
        print("\n✓ Test 4: Error log")
        error_log = driver.find_element(By.ID, "errorLog")
        print(f"  Error log text: {error_log.text[:100]}...")
        assert error_log is not None, "Error log should exist"
        
        # Test 5: Test Frank interaction
        print("\n✓ Test 5: Frank interaction")
        frank_input.send_keys("hi")
        send_button = driver.find_element(By.XPATH, "//button[text()='SEND']")
        send_button.click()
        time.sleep(1)
        
        output = driver.find_element(By.ID, "prologOutput")
        output_text = output.text
        print(f"  Frank response: {output_text[-100:]}")
        assert "Frank" in output_text, "Frank should respond"
        
        # Test 6: Check console errors
        print("\n✓ Test 6: Console errors")
        logs = driver.get_log('browser')
        errors = [log for log in logs if log['level'] == 'SEVERE']
        print(f"  Console errors: {len(errors)}")
        for error in errors[:3]:
            print(f"    - {error['message'][:80]}...")
        
        # Test 7: Check shards
        print("\n✓ Test 7: Monster Group shards")
        shards = driver.find_elements(By.CLASS_NAME, "shard")
        print(f"  Found {len(shards)} shards")
        assert len(shards) == 71, "Should have 71 shards"
        
        # Test 8: Click a shard
        print("\n✓ Test 8: Shard interaction")
        if shards:
            shards[0].click()
            time.sleep(1)
            # Check if modal appeared (proof modal)
            print("  Shard clicked successfully")
        
        # Test 9: Check Matrix effect
        print("\n✓ Test 9: Matrix background")
        canvas = driver.find_element(By.CLASS_NAME, "matrix-bg")
        print(f"  Matrix canvas found: {canvas is not None}")
        
        # Test 10: Check uptime
        print("\n✓ Test 10: Uptime counter")
        uptime = driver.find_element(By.ID, "uptime")
        uptime_text = uptime.text
        print(f"  Uptime: {uptime_text}")
        assert uptime_text, "Uptime should be displayed"
        
        print("\n" + "=" * 60)
        print("✅ All tests passed!")
        print(f"\nSummary:")
        print(f"  • Page loaded successfully")
        print(f"  • All UI elements present")
        print(f"  • Frank responds to queries")
        print(f"  • {len(errors)} console errors")
        print(f"  • 71 shards interactive")
        
        return True
        
    except Exception as e:
        print(f"\n❌ Test failed: {e}")
        
        # Take screenshot on failure
        driver.save_screenshot("dashboard_error.png")
        print("  Screenshot saved: dashboard_error.png")
        
        # Print page source for debugging
        with open("dashboard_source.html", "w") as f:
            f.write(driver.page_source)
        print("  Page source saved: dashboard_source.html")
        
        return False
        
    finally:
        driver.quit()

if __name__ == "__main__":
    success = test_dashboard()
    exit(0 if success else 1)

File: data/proofs/test_dashboard.rs
// test_dashboard.rs - Headless browser test in Rust

use headless_chrome::{Browser, LaunchOptions};
use std::time::Duration;
use std::thread;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("🔮 Testing zkPrologML Dashboard");
    println!("{}", "=".repeat(60));
    
    // Launch headless browser
    let browser = Browser::new(LaunchOptions {
        headless: true,
        ..Default::default()
    })?;
    
    let tab = browser.new_tab()?;
    
    // Load dashboard
    let url = "http://localhost/";
    println!("\n📥 Loading: {}", url);
    tab.navigate_to(url)?;
    tab.wait_until_navigated()?;
    
    thread::sleep(Duration::from_secs(2));
    
    // Test 1: Check title
    println!("\n✓ Test 1: Page title");
    let title = tab.get_title()?;
    println!("  Title: {}", title);
    assert!(title.contains("zkPrologML"), "Title should contain 'zkPrologML'");
    
    // Test 2: Check stats cards
    println!("\n✓ Test 2: Stats cards");
    let stats = tab.find_elements(".card")?;
    println!("  Found {} stat cards", stats.len());
    assert!(stats.len() >= 6, "Should have at least 6 stat cards");
    
    // Test 3: Check Frank input
    println!("\n✓ Test 3: Frank assistant");
    let frank_input = tab.find_element("#prologInput")?;
    let placeholder = frank_input.get_attribute_value("placeholder")?;
    println!("  Frank placeholder: {:?}", placeholder);
    
    // Test 4: Check error log
    println!("\n✓ Test 4: Error log");
    let error_log = tab.find_element("#errorLog")?;
    let error_text = error_log.get_inner_text()?;
    println!("  Error log: {}...", &error_text[..error_text.len().min(50)]);
    
    // Test 5: Test Frank interaction
    println!("\n✓ Test 5: Frank interaction");
    frank_input.type_into("hi")?;
    let send_button = tab.find_element("button")?;
    send_button.click()?;
    thread::sleep(Duration::from_secs(1));
    
    let output = tab.find_element("#prologOutput")?;
    let output_text = output.get_inner_text()?;
    println!("  Frank response: {}...", &output_text[output_text.len().saturating_sub(100)..]);
    assert!(output_text.contains("Frank"), "Frank should respond");
    
    // Test 6: Check shards
    println!("\n✓ Test 6: Monster Group shards");
    let shards = tab.find_elements(".shard")?;
    println!("  Found {} shards", shards.len());
    assert_eq!(shards.len(), 71, "Should have 71 shards");
    
    // Test 7: Check uptime
    println!("\n✓ Test 7: Uptime counter");
    let uptime = tab.find_element("#uptime")?;
    let uptime_text = uptime.get_inner_text()?;
    println!("  Uptime: {}", uptime_text);
    
    // Test 8: Take screenshot
    println!("\n✓ Test 8: Screenshot");
    let screenshot = tab.capture_screenshot(
        headless_chrome::protocol::cdp::Page::CaptureScreenshotFormatOption::Png,
        None,
        None,
        true,
    )?;
    std::fs::write("dashboard_test.png", screenshot)?;
    println!("  Screenshot saved: dashboard_test.png");
    
    println!("\n{}", "=".repeat(60));
    println!("✅ All tests passed!");
    println!("\nSummary:");
    println!("  • Page loaded successfully");
    println!("  • All UI elements present");
    println!("  • Frank responds to queries");
    println!("  • 71 shards interactive");
    println!("  • Screenshot captured");
    
    Ok(())
}
